extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -1013,6 +1051,93 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
     return ['https://partner.googleadservices.com',
       'https://tpc.googlesyndication.com'];
   }
+
+  /** @override */
+  getNonAmpCreativeRenderingMethod(headerValue) {
+    return this.isFluid_ ? XORIGIN_MODE.SAFEFRAME :
+        super.getNonAmpCreativeRenderingMethod(headerValue);
+  }
+
+  /** @override */
+  getAdditionalContextMetadata() {
+    const attributes = /** @type {!JsonObject} */ ({});
+    if (this.isFluid_) {
+      attributes['uid'] = 1;
+      attributes['hostPeerName'] = this.win.location.origin;
+      // The initial geometry isn't used for anything important, but it is
+      // expected, so we pass this string with all zero values.
+      attributes['initialGeometry'] = JSON.stringify(
+          /** @type {!JsonObject} */ ({
+            'windowCoords_t': 0,
+            'windowCoords_r': 0,
+            'windowCoords_b': 0,
+            'windowCoords_l': 0,
+            'frameCoords_t': 0,
+            'frameCoords_r': 0,
+            'frameCoords_b': 0,
+            'frameCoords_l': 0,
+            'styleZIndex': 'auto',
+            'allowedExpansion_t': 0,
+            'allowedExpansion_r': 0,
+            'allowedExpansion_b': 0,
+            'allowedExpansion_l': 0,
+            'xInView': 0,
+            'yInView': 0,
+          }));
+      attributes['permissions'] = JSON.stringify(
+          /** @type {!JsonObject} */ ({
+            expandByOverlay: false,
+            expandByPush: false,
+            readCookie: false,
+            writeCookie: false,
+          }));
+      attributes['metadata'] = JSON.stringify(
+          /** @type {!JsonObject} */ ({
+            shared: {
+              'sf_ver': this.safeframeVersion,
+              'ck_on': 1,
+              'flash_ver': '26.0.0',
+            },
+          }));
+      attributes['reportCreativeGeometry'] = true;
+      attributes['isDifferentSourceWindow'] = false;
+      attributes['sentinel'] = this.sentinel;
+    }
+    return attributes;
+  }
+
+  /** @override */
+  getXdomainCreativeFrameMessageListeners() {
+    return !this.isFluid_ ? {} : {
+      'creative_geometry_update': data => {
+        // The first creative_geometry_update message will contain bad
+        // geometric data, as it will have been computed using the initial,
+        // incorrect, iframe style. We use this first message as a
+        // triggering point to restyle the iframe, which will in turn cause
+        // safeframe to send another creative_geometry_update message, this
+        // time containing the correct information.
+        const styleString = this.iframe.getAttribute('style');
+        if (/width: 0px/.test(styleString) &&
+            /height: 0px/.test(styleString)) {
+          setStyles(this.iframe, {
+            width: '100%',
+            height: '100%',
```

Comment:
What does `height = 100%` mean?

Reply:
In this context, it tells the iframe to take up the entire height of the ad slot. Without that line, the iframe has height 0.